[WIP] Buffers panel highlight current

MaxGyver83 opened the pull request:
# Description

While `switch-buffer` is very efficient for finding a specific buffer among many, it requires that you remember what to search for. Having a list of open buffers always visible (as p.e. in Firefox) helps to keep an overview. Nyxt's buffer panel is very similar to a tab bar in other browsers.

In this pull request, I try to make the buffer panel even a bit more useful.

## To Do:

* ✔ Make the reload button stand out by paining it green and moving it above the header.
* ✔ Highlight the current buffer by painting its button blue (currently only updated when the buffer is changed using the buffer panel).
* ✘ Reload the buffer panel via `window-set-buffer-hook`.

### Before:

![buffers-panel-old](https://github.com/atlas-engineer/nyxt/assets/13970628/4d2117bc-40e9-41f7-a4c4-9ff1adced010)

### After:

![buffers-panel-new](https://github.com/atlas-engineer/nyxt/assets/13970628/64d3f992-c37a-442a-99e6-81affaa4c6e4)

# Discussion

Please tell me if such a changed is welcome at all. If yes, you might prefer a different design. I actually want to use the default `prompt-buffer` highlight color instead of a simple blue but I couldn't manage to select this theme color. I also don't yet understand how to set a hook.

One detail: In line 100, I create either a regular or a blue `:nbutton` depending on the `highlight` argument. Actually, I would prefer to write less duplicate code but this does not work:

```lisp
(:nbutton :text (title buffer)
                  :buffer panel-buffer
                  @(if highlight '(:style "background-color:blue;"))
                 ...
```

I don't know why. Probably easy to solve but I'm a Common Lisp beginner.

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.

- [x] Git hygiene:
  - I have pulled from master before submitting this PR
  - There are no merge conflicts.
- [ ] My code follows the style guidelines for Common Lisp code. See:
  - [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf)
  - [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml)
- [x] I have performed a self-review of my own code.
- [ ] My code has been reviewed by at least one peer.  (The peer review to approve a PR counts.  The reviewer must download and test the code.)
- [ ] Documentation:
  - All my code has docstrings and `:documentation`s written in the aforementioned style.  (It's OK to skip the docstring for really trivial parts.)
  - I have updated the existing documentation to match my changes.
  - I have commented my code in hard-to-understand areas.
  - I have updated the `changelog.lisp` with my changes if it's anything user-facing (new features, important bug fix, compatibility breakage).
    - Changelog update should be a separate commit.
  - I have added a `migration.lisp` entry for all compatibility-breaking changes.
  - (If this changes something about the features showcased on Nyxt website) I have these changes described in the new/existing article at Nyxt website or will notify one of maintainters to do so.
- [ ] Compilation and tests:
  - My changes generate no new warnings.
  - I have added tests that prove my fix is effective or that my feature works.  (If possible.)
  - I ran the tests locally (`(asdf:test-system :nyxt)` and `(asdf:test-system :nyxt/gi-gtk)`) and they pass.

aartaka commented on source/mode/buffer-listing.lisp:
```diff
@@ -88,36 +88,43 @@ With LINEAR-VIEW-P, list buffers linearly instead."
-                   :buffer panel-buffer
-                   `(nyxt::switch-buffer :buffer ,buffer))))))
+             (:p (if highlight
+                     (:nbutton :text (title buffer)
```
You can rely on Spinneret expanding style attributes from Lisp expressions:
``` lisp
(:nbutton
  :text (title buffer)
  :buffer panel-buffer
  :style (when highlight "background-color:blue;")
  ...)
```

Spinneret as a library is quite smart, so you can pass regular Lisp expressions to it and be (99%) sure that they will be interpreted properly.

MaxGyver83 replied:
This works like a charm! Thank you. I'll change it.

MaxGyver83 replied:
Applied in 6ebccfc5.

aartaka commented on source/mode/buffer-listing.lisp:
```diff
@@ -88,36 +88,43 @@ With LINEAR-VIEW-P, list buffers linearly instead."
       (:body
-       (:h1 "Buffers")
        (:nbutton :text "Update ↺"
+         :style "background-color:green;"
```
You can use style classes exported by Nyxt to color things in a more consistent and theme-dependent shades. Like:
``` lisp
;; For green.
(:nbutton
  :class "success"
  ...)
;; For blue.
(:nbutton
  :class "action"
  ...)
```
----------------------------------------------------

Also note that inline styles are hard to maintain, so you better put the styles into the `style` slot of the respective mode (`buffer-listing-mode` here) or use the built-in style classes, as I've suggested.

MaxGyver83 replied:
I have tried this. It works for `:class "action"` but not for `:class "success"`. Where is the according CSS defined?

MaxGyver83 replied:
OK, I think I have found it:
<https://github.com/MaxGyver83/nyxt/blob/be7ca1c5cabde6eb857581a35b0734a1694c0acf/source/prompt-buffer.lisp#L136-L138>

aartaka replied:
No, prompt buffer is not related to panel buffer. What you need is buffer.lisp, and `style` slot inside `panel-buffer`. 

But there were no proper CSS classes there, so I added them in 8cb76d447c6ed5e6e26b8b561763c1f95532fb83. Rebase over master and try again. Should work this time.

MaxGyver83 replied:
Now, green and blue are use as foreground colors:
![buffers-panel-new2](https://github.com/atlas-engineer/nyxt/assets/13970628/4d4d94da-3a63-47b8-ad03-9469df29425b)
Was this intended? I think the contrast of blue on grey could be better.

aartaka replied:
Hmmmmm, that's not right. And I know why. Fixed in 08573888a2cda3826faaa57719733ebdbe275241. Thanks for highlighting (pun not intended) it!

MaxGyver83 replied:
:-) Thank you for fixing it. It looks better now! I have updated the screenshot in the pull request description.

aartaka commented on source/mode/buffer-listing.lisp:
```diff
@@ -88,36 +88,43 @@ With LINEAR-VIEW-P, list buffers linearly instead."
-             collect (:raw (buffer-markup buffer)))))))
+             collect (:raw (buffer-markup buffer (eq buffer (current-buffer)))))))))
+
+(defun reload-panel-buffer (panel-buffer)
```
This might go into the body of the panel command as:
``` lisp
(reload-buffer panel-buffer)
```

So no need for a separate function that works out of context.

aartaka replied:
Oh, right. I haven't taken the window hook into account. For that, a separate function might be necessary. But here, in buttons, you can benefit from the panel buffer accessibility.

MaxGyver83 replied:
I have added the window hook in commit bec09b2b. Now I call `reload-panel-buffer` in the hook. Is it correct to call this top-level function also in the buffers panel? Or should I create a `flet` expression nevertheless?

aartaka commented on source/mode/buffer-listing.lisp:
```diff
@@ -88,36 +88,43 @@ With LINEAR-VIEW-P, list buffers linearly instead."
+                      `(nyxt::switch-buffer :buffer ,buffer))
+                     (:nbutton :text (title buffer)
+                      :buffer panel-buffer
+                      `(progn
```
You can pass several expressions as `:nbutton` body:
``` lisp
(:nbutton
  ...
  `(nyxt::switch-buffer :buffer ,buffer)
  `(reload-buffer ,panel-buffer))
```

MaxGyver83 replied:
Applied in commit 6ebccfc51.

aartaka reviewed (commented):
I like the idea! I've left some comments, mostly cosmetic and helping with Spinneret and our custom tags, like `:nbutton`.

aartaka commented:
>I have tried this. It works for `:class "action"` but not for `:class "success"`. Where is the according CSS defined?

It's buffer.lisp, line 167 on master.

MaxGyver83 commented:
So far, my best try  to add `reload-panel-buffer` to the `window-set-buffer-hook`:

```lisp
(defun buffers-panel-handler (window buffer)
  (reload-panel-buffer (first (nyxt::panel-buffers (current-window)))))

(define-configuration window
  ((window-set-buffer-hook
    (hooks:add-hook %slot-default% 'buffers-panel-handler))))
```

AFAIU, this warning lets the compilation fail:

```
Compilation warning: The variable WINDOW is defined but never used.
```

When I remove the function arguments `window` and `buffer`, I get a different error.

aartaka commented:
>Compilation warning: The variable WINDOW is defined but never used

It's just a compiler being picky. Add (declare (ignore window)) to the handler right after argument list and it'll work.

aartaka commented on source/mode/buffer-listing.lisp:
```diff
@@ -88,36 +88,47 @@ With LINEAR-VIEW-P, list buffers linearly instead."
-            :key (compose
-                  #'render-url #'url))))
+         :class "success"
+         `(reload-panel-buffer ,panel-buffer))
```
This still feels like an unnecessary indirection. Can't you use a raw 
``` lisp
(reload-buffer ,panel-bufer)
```
here? I might be out of context, so correct me if I'm missing something.

MaxGyver83 replied:
This seems to work. At least I don't get an error. Pressing the reload button has no visible effect. I'm wondering if it's needed at all with this pull request. The buffers panel is reloaded via `window-set-buffer-hook` whenever the user switches to another buffer. And when a buffer is closed, the previous(?) buffer is loaded automatically which triggers the hook, too. Is there still a use case which requires a manual reload?

Also, if `` `(reload-buffer ,panel-bufer)`` works, why was `` `(reload-buffer (find ...))`` used initially? Does it make a difference when there are multiple panel buffers!?

MaxGyver83 replied:
Do you mean to keep `(reload-buffer (find ...))` or to keep `(reload-panel-buffer ,panel-buffer)` (which is a wrapper around the former to prevent duplicate code)?

Maybe the first question is if I can add a hook that calls `reload-panel-buffer` at all!? @aartaka [said](https://github.com/atlas-engineer/nyxt/pull/3199/files#r1344010508) that I can't use `define-configuration` in the codebase. Is there a different way to add a hook directly in the code?

aartaka commented on source/mode/buffer-listing.lisp:
```diff
@@ -88,36 +88,47 @@ With LINEAR-VIEW-P, list buffers linearly instead."
+    (when panel-buffer
+      (reload-panel-buffer panel-buffer))))
+
+(define-configuration window
```
We had an issue about actually using `define-configuration` in our codebase: #2877. Bu it needs a new architecture to work properly. Until then, we leave `define-configuration` as a strictly user config construct.

How about dropping this `define-configuration`, using `reload-buffer` in `buffers-panel`, and removing `buffers-panel-handler`&`reload-panel-buffer`? We lose the updates this way, but it's more reliable this way. You can leave a `TODO`/`FIXME` in the code with the mention of live updates.

Sorry if it's a bit messy and non-interactive—need a bit of refactor to make it all work properly :sweat_smile:

MaxGyver83 replied:
I don't yet understand why `define-configuration` isn't allowed. (I only had a quick look at #2877.)

Don't you think it's pointless highlighting the current buffer when it doesn't get updated automatically!? This would mean that in some cases (p.e. when using `switch-buffer`) a wrong buffer remains highlighted which is rather confusing. Maybe it's better to have no buffer highlighted than a wrong one.

aartaka commented:
@aadcg, take a look. Most of my programming-related questions are resolved, so what's left is the feature assessment.

MaxGyver83 commented on source/mode/buffer-listing.lisp:
```diff
@@ -88,36 +88,47 @@ With LINEAR-VIEW-P, list buffers linearly instead."
     (panel-buffer "*Buffers panel*")
   "Display a list of buffers with easy switching."
-  (flet ((buffer-markup (buffer)
+  (flet ((buffer-markup (buffer &optional highlight)
```
I have changed it: 62559ce1e4d70c46767fb5fc81a2defff78c0d29.

MaxGyver83 commented:
> Steps to reproduce:
> 
>     * `(nyxt:start :urls '("https://en.wikipedia.org/wiki/Tomato") :failsafe t)`
> 
>     * Invoke command `buffers-panel`
> 
>     * Invoke command `set-url-new-buffer`
> 
>     * Notice the behavior above.

True. It looks like my testing was too narrow. I have committed two changes:

* Print "Loading..." instead of a tiny empty button when the title of a new buffer is not yet known.
* I have added a `buffer-loaded-hook`. I think both hooks are necessary.

I'm still trying to add the `window-set-buffer-hook` directly in `window.lisp` in `initialize-instance`.

MaxGyver83 commented:
> @MaxGyver83 would you like to make the changes yourself or shall I do it? Thanks.

I have committed your suggested change, updated the changelog and rebased.
